### PR TITLE
raftstore: Implement coprocessor observer pre_exec_admin(query)

### PR DESCRIPTION
**Pull request by CalvinNeo**
Signed-off-by: CalvinNeo <calvinneo1995@gmail.com>

### What is changed and how it works?

Issue Number: ref #12849

What's Changed:

```commit-message
Support new observers pre_exec_admin(query).
```

Note that when one of pre_exec_admin(query) observers returns true, raftstore will skip exec this cmd, but still reckon as applied.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -5491,6 +5525,88 @@ mod tests {
         });
     }
 
+    #[test]
+    fn test_exec_observer() {
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/fsm/apply.rs:5529]: please write some comments about what's the test points of this test and explain which parts of the new code are tested.

File: components/raftstore/src/store/fsm/apply.rs [hunk 2]
```diff
@@ -1242,32 +1242,44 @@ where
...
         // E.g. `RaftApplyState` must not be changed.
-        let (resp, exec_result) = match self.exec_raft_cmd(ctx, req) {
-            Ok(a) => {
-                ctx.kv_wb_mut().pop_save_point().unwrap();
-                if req.has_admin_request() {
-                    origin_epoch = Some(self.region.get_region_epoch().clone());
-                }
-                a
+
+        let (resp, exec_result) = if ctx.host.pre_exec(&self.region, req) {
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/fsm/apply.rs:1246]: please add a debug log to indicate the execution is skipped?

Comment (tonyxuqqi) [reply components/raftstore/src/store/fsm/apply.rs:1246]: or add a metrics  for pre_exec return true

**Comment by tonyxuqqi** [comment]
Generally LGTM

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore/src/store/fsm/apply.rs [hunk 3]
```diff
@@ -1218,6 +1218,20 @@ where
         exec_result
     }
 
+    fn need_pre_exec(&self, req: &RaftCmdRequest) -> bool {
+        if req.has_admin_request() {
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:1222]: Why? I think all commands need to exec `pre_exec`.

Comment (CalvinNeo) [reply components/raftstore/src/store/fsm/apply.rs:1222]: Yes, `pre_exec` is designed for all commands. If one want to pre_exec some kind of commands, he will register those in `need_pre_exec`.
The `need_pre_exec` is like a highlight for developers and reviewers. If a command like `BatchSplit` needs to be `pre_exec`, and thus be potentially skipped execution, we'd better think out why.

Comment (CalvinNeo) [reply components/raftstore/src/store/fsm/apply.rs:1222]: Maybe I can pre_exec for every command, and skip when need_pre_exec is true.

Comment (BusyJay) [reply components/raftstore/src/store/fsm/apply.rs:1222]: No, this is just a trap for observer authors.

Comment (CalvinNeo) [reply components/raftstore/src/store/fsm/apply.rs:1222]: Then we just remove `need_pre_exec`?

File: components/raftstore/src/store/fsm/apply.rs [hunk 4]
```diff
@@ -1242,32 +1242,44 @@ where
...
         // Remember if the raft cmd fails to be applied, it must have no side effects.
         // E.g. `RaftApplyState` must not be changed.
-        let (resp, exec_result) = match self.exec_raft_cmd(ctx, req) {
-            Ok(a) => {
-                ctx.kv_wb_mut().pop_save_point().unwrap();
-                if req.has_admin_request() {
-                    origin_epoch = Some(self.region.get_region_epoch().clone());
-                }
-                a
+
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:1245]: L1239 ~ L1241 should be moved to L1255.

**Review by BusyJay — APPROVED** [review]

**Review by breezewish — COMMENTED** [review]
/merge

**Pull request merged**